design/design-doc-resource-configurations.md

```diff
@@ -0,0 +1,838 @@
+design. Defining a new Kubernetes resource kind typically involves authoring a
+[Custom Resource Definition] (CRD). A method to simplify and constrain this
+process is desirable, and may be explored in future.
+
+## Proposal
+
+This document proposes Crossplane introduce two new kinds:
+`ResourceConfiguration` and `ResourceConfigurationClass`. These new kinds would
+exist in the `infrastructure.crossplane.io` API group. Note that in this context
+"configuration" should be interpreted as in the dictionary definition of the
+word:
+
+> an arrangement of elements in a particular form, figure, or combination.
+
+A `ResourceConfiguration` represents a configuration of resources; a composite
+of the primitive managed resources supported by each infrastructure provider. A
+`ResourceConfiguration` groups managed resources into a single binding target
+for a resource claim, and defines which resource claim kinds may bind to an
+unbound configuration of resources. A `ResourceConfigurationClass` determines
+which primitive managed resources should be dynamically provisioned to satisfy a
+particular kind of resource claim, including specifying how resource claim
+fields should relate to the dynamically provisioned managed resources.
+
+### Example: Azure MySQL Server with Virtual Network Rule
+
+The below example uses a user defined resource claim (`MySQLInstance`) along
+with a `ResourceConfigurationClass` to produce a `ResourceConfiguration`
+consisting of an Azure MySQL server and a virtual network rule.
+
+```yaml
+---
+# Note that this MySQLInstance is in the database.example.org API group (not
+# database.crossplane.io). It includes fields not supported by the contemporary
+# MySQLInstance.database.crossplane.io claim kind.
+apiVersion: database.example.org/v1alpha1
+kind: MySQLInstance
+metadata:
+  namespace: default
+  name: sql
+  annotations:
+    # The external name annotation is copied to all managed resources that are
+    # dynamically provisioned by this claim.
+    crossplane.io/external-name: very-private-database
+spec:
+  engineVersion: "5.7"
+  storageGB: 10
+  region: us-west
+  # Resource claim custom resources must support Crossplane's claim binding
+  # metadata; they must have a writeConnectionSecretToRef, classRef,
+  # resourceRef, bindingPhase, and status conditions.
+  writeConnectionSecretToRef:
+    name: sql
+  classRef:
+    apiVersion: infrastructure.crossplane.io/v1alpha1
+    kind: ResourceConfigurationClass
+    name: private-mysql-server
+  resourceRef:
+    apiVersion: infrastructure.crossplane.io/v1alpha1
+    kind: ResourceConfiguration
+    name: default-sql-gdm0w
+---
+apiVersion: infrastructure.crossplane.io/v1alpha1
+kind: ResourceConfigurationClass
+metadata:
+  name: private-mysql-server
+spec:
+  # Each ResourceConfigurationClass defines how resources are dynamically
+  # provisioned for exactly one kind of resource claim; there is a one-to-one
+  # claim-kind-to-configuration-class-instance relationship.
```

Comment:
Note to self: this is a lie. There's a one-to-many claim-kind-to-configuration-class-instance relationship.